go/vendor/github.com/keybase/go-updater/watchdog/watchdog.go

```diff
@@ -57,15 +60,62 @@ func Watch(programs []Program, restartDelay time.Duration, log Log) error {
 	return nil
 }
 
+// terminate will send a kill signal to the program by finding its pid from the
+// path to its executable. If the program has specified a pidfile, this function
+// will open up that file and see if the pid in there matches the pid that was just
+// terminated. If so, this function will delete the file. It is much better for
+// programs to manage their own pidfiles, but kill signals cannot always be caught
+// and handled gracefully (i.e. in windows).
+func (p Program) terminate(log Log) {
+	log.Infof("Terminating %s", p.Path)
+	thisProcess := os.Getpid()
+	matcher := process.NewMatcher(p.Path, process.PathEqual, log)
+	matcher.ExceptPID(thisProcess)
+	// this logic also exists in the updater, so if you want to change it, look there too.
+	terminatedPids := process.TerminateAll(matcher, time.Second, log)
+	if len(terminatedPids) == 0 {
+		// nothing terminated probably because nothing was running
+		return
+	}
+	// if there was a pidfile, it might not have been cleaned up, so let's take a look
+	if p.PidFile == "" {
+		// nothing more to do
+		return
+	}
+	time.Sleep(500 * time.Millisecond) // give the program a chance to clean itself up
```

Comment:
maybe we could try looping here few times (so at most few seconds total) to increase our chance of the lock being released and us cleaning up after it

Reply:
sure. that makes sense.